src/main/java/com/commercetools/sync/commons/utils/QueryAll.java

```diff
@@ -1,166 +1,180 @@
-     * @return a future containing a list of results of the callback on each page.
+     * @param pageMapper the function to apply on each fetched page of the result space.
+     * @return a future containing a list of mapped results of type {@code S}, after the function applied all the pages.
      */
     @Nonnull
-    <S> CompletionStage<List<S>> run(final SphereClient client, final Function<List<T>, S> callback) {
-        return queryPage(client, 0).thenCompose(result -> {
-            final List<CompletableFuture<S>> futureResults = new ArrayList<>();
-
-            final S callbackResult = callback.apply(result.getResults());
-            final List<CompletableFuture<S>> nextPagesCallbackResults =
-                queryNextPages(client, result.getTotal(), callback);
-
-            futureResults.add(completedFuture(callbackResult));
-            futureResults.addAll(nextPagesCallbackResults);
-            return transformListOfFuturesToFutureOfLists(futureResults);
-        });
+    CompletionStage<List<S>> run(@Nonnull final Function<List<T>, S> pageMapper) {
+        this.pageMapper = pageMapper;
+        final CompletionStage<PagedQueryResult<T>> firstPage = queryPage();
+        return queryNextPages(firstPage)
+            .thenApply(nextPage -> nextPage.mappedResultsTillNow);
     }
 
     /**
-     * Given a {@link Consumer}, this method applies this consumer on each page of the results from
-     * {@link this} instance's {@code baseQuery}.
+     * Given a {@link Consumer} to a page of resources of type {@code T}, this method sets this instance's
+     * {@code pageConsumer} to the supplied value, then it makes requests to fetch the entire result space of the
+     * resource {@code T} on CTP, while accepting the consumer on each fetched page.
      *
-     * @param client   the CTP client that the query is run on.
-     * @param consumer the consumer that gets called on each page of results.
-     * @return an empty future.
+     * @param pageConsumer the consumer to accept on each fetched page of the result space.
+     * @return a future containing void after the consumer accepted all the pages.
      */
     @Nonnull
-    CompletionStage<Void> run(final SphereClient client, final Consumer<List<T>> consumer) {
-        return queryPage(client, 0).thenCompose(result -> {
-            final List<CompletableFuture<Void>> futureResults = new ArrayList<>();
-            consumer.accept(result.getResults());
-            futureResults.addAll(queryNextPages(client, result.getTotal(), consumer));
-            return CompletableFuture.allOf(futureResults.toArray(new CompletableFuture[futureResults.size()]));
-        });
+    CompletionStage<Void> run(@Nonnull final Consumer<List<T>> pageConsumer) {
+        this.pageConsumer = pageConsumer;
+        final CompletionStage<PagedQueryResult<T>> firstPage = queryPage();
+        return queryNextPages(firstPage).thenAccept(result -> { });
```

Comment:
> result

I recommend to call such variables `ignore` so it is explicitly visible nothing is missed in the implementation.

Reply:
2638c02